### ISSUE TITLE
Add `fileCreated` to the File Handling interface?

### ISSUE THREAD

**Issue by dbrnz** [opencor/opencor#1028]
Although `fileReloaded()` is called when a new CellML file is saved, it only does so if the contents are valid. No notification is given if the file contains invalid CellML and the user decides to save the file as text.

How about calling a new method, `fileCreated(const QString &pFileName)`, when a file is saved for the first time? This should also be called when "Save As" is used to save a copy of a file.

**Comment by agarny** [comment]
What are you trying to do exactly?

If you want to create a new file, you should do this through the file manager instance. For example, in the CellML Support plugin, if the user wants to create a new CellML file (via the menu), then [`CellMLSupportPlugin::newCellmlFile()`](https://github.com/opencor/opencor/blob/572c7fe1cfd0f0cba67f8183f26d96e91cd5cb8a/src/plugins/support/CellMLSupport/src/cellmlsupportplugin.cpp#L205) is eventually called, and there you can see how we can actually create a new file in OpenCOR (here, a new CellML file with some initial contents). From there, you can see that [`FileManager::create()`](https://github.com/opencor/opencor/blob/572c7fe1cfd0f0cba67f8183f26d96e91cd5cb8a/src/plugins/miscellaneous/Core/src/filemanager.cpp#L482) emits the `fileCreated` signal, which you can easily handle, if needed.

**Comment by dbrnz** [comment]
I'm wanting to detect that new files have been created and existing files changed in a workspace folder.

For plugins using the file manager interface, `fileReloaded()` is called when the editor saves an existing file; however it is not called if a new file contains invalid CellML and the user chooses to save the file as text. For example, methods and signals when creating and saving a new file containing valid CellML are:

```
fileCreated:  "/var/folders/7t/_0gntlnx7x75wcpl69zv5yyh0000gp/T/K56833.tmp"
fileRenamed:  "/private/var/folders/7t/_0gntlnx7x75wcpl69zv5yyh0000gp/T/K56833.tmp"
   to:  "/Users/dave/Documents/OpenCOR/Workspaces/A New Workspace/File #1.cellml"
fileReloaded:  "/Users/dave/Documents/OpenCOR/Workspaces/A New Workspace/File #1.cellml"
```

When creating and saving a new file with invalid CellML the only signal emitted is fileCreated; no methods are called:

```
fileCreated:  "/var/folders/7t/_0gntlnx7x75wcpl69zv5yyh0000gp/T/V56833.tmp"
```

If the editor window is closed after saving the file as text, and the file reopened for editing, then file save results in:

```
fileReloaded:  "/Users/dave/Documents/OpenCOR/Workspaces/A New Workspace/File #2.txt"
```

**Comment by agarny** [comment]
Ok, I am starting to appreciate what you are trying to do.

I have yet to think properly about it, but you can see [here](https://github.com/opencor/opencor/blob/2be4d2581caa24ff797c0e93f772a2e509b6684f/src/plugins/miscellaneous/Core/src/filemanager.h#L159) that `FileManager` emits some signals that it considers useful. Now, the file handling interface doesn't have matching methods for all of those signals, but I guess that we could, in principle, do that. Thoughts?

**Comment by agarny** [comment]
@dbrnz, any thoughts on the above before I give it some serious thoughts myself?

**Comment by dbrnz** [comment]
The signals emitted by `FileManager` along with the methods provided by the file handling interface are just fine for what I want (namely to be informed about files created and modified by OpenCOR's editing plugins) apart from the fact that when a newly created file doesn't contain valid CellML no notification is given of the name that the file is saved as...

Operation | File contains valid CellML | File contains invalid CellML
-----------|--------------------------|----------------------------
Save new file: | Renamed signal emitted and reloaded method called | *Nothing emitted nor called*
Save modified file: | Reloaded method called | Reloaded method called

**Comment by agarny** [comment]
@dbrnz, sorry for the delay in getting back to you. I am having another look at your issue and I must confess that the table in your latest message confuses me.

I mean, if, on OS X, I create a new CellML file (resulting in `FileHandlingInterface::fileOpened()` and `FileHandlingInterface::filePermissionsChanged()` being called) and edit it using the `CellML Text` view and end up with:
 - a valid CellML file, then I can save it (resulting in `FileHandlingInterface::fileRenamed()`, `FileHandlingInterface::fileReloaded()` and `FileHandlingInterface::filePermissionsChanged()` being called).
 - an invalid CellML file, then I cannot save it (and no `FileHandlingInterface` method gets called).

Now, if I open an existing CellML file (resulting in `FileHandlingInterface::fileOpened()` and `FileHandlingInterface::filePermissionsChanged()` being called) and edit it using the `CellML Text` view and end up with:
 - a valid CellML file, then I can save it (resulting in `FileHandlingInterface::fileReloaded()` being called).
 - an invalid CellML file, then I cannot save it (and no `FileHandlingInterface` method gets called).

In other words, no matter what, if I try to save an invalid CellML file using the `CellML Text` view, I won't be able to do so and no `FileHandlingInterface` method will get called (both as expected).

However, when you create a new CellML file or open an existing one, you can see that both `FileHandlingInterface::fileOpened()` and `FileHandlingInterface::filePermissionsChanged()` are called, so maybe that's enough for what you want?

If I missed something, then please provide me with the exact steps of what you are trying to do so that I can see where things don't go as you would expect.

**Comment by dbrnz** [comment]
1. Use "File/New/File" to open an editor tab for a new file.
2. Enter text that isn't valid CellML.
3. Use "File/Save", specify a file name, and "Save".
4. Respond "Yes" to the dialog offering to save the file as text.
5. "Save" the file.

This sequence creates a text file on disk with neither `FileHandlingInterface::fileRenamed()` nor `FileHandlingInterface::fileReloaded()` being called.

Next:

1. Close the editing tab, responding "No" to the dialog offering to save the file.
2. Use "File/Open" to open the file.
3. Make changes (still invalid CellML) and save with "File/Save".
 
This updates the file on disk with `FileHandlingInterface::fileReloaded()` **being called**.

I'm hoping that a successful save of a new file can always result in both `fileRenamed()` and `fileReloaded()` being called, just as a successful save of an existing file results in `fileReloaded()` being called.

**Comment by agarny** [comment]
> 1. Use "File/New/File" to open an editor tab for a new file.
2. Enter text that isn't valid CellML.
3. Use "File/Save", specify a file name, and "Save".
4. Respond "Yes" to the dialog offering to save the file as text.
5. "Save" the file.

> This sequence creates a text file on disk with neither `FileHandlingInterface::fileRenamed()` nor `FileHandlingInterface::fileReloaded()` being called.

Just to be 100% clear, I am assuming that you are performing those steps using the `CellML Text` view (based on your step 4). If so, then what you have observed is completely normal.

Indeed, that feature (to be able to save a malformed CellML file as a text file; see issue #998) is purely for user convenience. So, such a file is not and was never meant to be tracked by OpenCOR, hence no `FileHandlingInterface` method is called for that file (as is the case when you export some simulation data to CSV/BioSignalML, or a simulation to a SED-ML file or a COMBINE archive). In fact, consider that text file as being simply some kind of a 'scrapbook'. Nothing more nothing less.

So, personally, I would probably 'manually' track changes to a workspace, not least because it might very well be the case that someone creates a file in a workspace and this outside of OpenCOR. I guess that's probably something that might be handled by some kind of a Git support plugin?

> 1. Close the editing tab, responding "No" to the dialog offering to save the file.
2. Use "File/Open" to open the file.
3. Make changes (still invalid CellML) and save with "File/Save".
 
> This updates the file on disk with `FileHandlingInterface::fileReloaded()` **being called**.

> I'm hoping that a successful save of a new file can always result in both `fileRenamed()` and `fileReloaded()` being called, just as a successful save of an existing file results in `fileReloaded()` being called.

Based on what I mentioned above, this next set of steps doesn't make sense as such. Indeed, the file that was saved earlier is a pure text file, not a CellML file.

Hmm, maybe this should be made clearer when offering the user to save a malformed CellML file to a text file?

**Comment by dbrnz** [comment]
> Just to be 100% clear, I am assuming that you are performing those steps using the CellML Text view (based on your step 4). If so, then what you have observed is completely normal.

Yes, I'm using the CellML Text View.

> So, personally, I would probably 'manually' track changes to a workspace, not least because it might very well be the case that someone creates a file in a workspace and this outside of OpenCOR. I guess that's probably something that might be handled by some kind of a Git support plugin?

I was hoping that the File Handling Interface would signal details of all new and modified (and deleted?) files in a workspace when OpenCOR is used to make changes. My preference is to keep things simple (i.e. hear about file system changes directly from OpenCOR) and not use things such as `QFileSystemWatcher` (potential file handle limitations), a timed refresh, nor something like [fswatch](https://github.com/emcrisostomo/fswatch).

If a user changes workspace files outside of OpenCOR then it is up to them to manually refresh things (say by opening and closing the workspace in the tree view).

> Based on what I mentioned above, this next set of steps doesn't make sense as such. Indeed, the file that was saved earlier is a pure text file, not a CellML file.

> Hmm, maybe this should be made clearer when offering the user to save a malformed CellML file to a text file?

OpenCOR currently has three options under "File/New" — "File", "CellML File" and "SED-ML File". If I user chooses to create a file of a particular type then should they be held to that type? What has confused me is that "File/Open" will happily open and save text files without complaint (including C++ sources) calling appropriate File Handling Interface methods as it does so.

How about having the `FileManager` emit signals regardless of a file's type but only call interface methods if the contents are of a known type and valid (for the type)??

**Comment by agarny** [comment]
> I was hoping that the File Handling Interface would signal details of all new and modified (and deleted?) files in a workspace when OpenCOR is used to make changes. My preference is to keep things simple (i.e. hear about file system changes directly from OpenCOR) and not use things such as `QFileSystemWatcher` (potential file handle limitations), a timed refresh, nor something like [fswatch](https://github.com/emcrisostomo/fswatch).

`FileHandlingInterface` has always been for handling files that have a corresponding file tab in the GUI. So, all other files, like the ones I mentioned in my previous message, are unknown to OpenCOR's file manager and therefore to `FileHandlingInterface`.

> OpenCOR currently has three options under "File/New" — "File", "CellML File" and "SED-ML File". If I user chooses to create a file of a particular type then should they be held to that type? What has confused me is that "File/Open" will happily open and save text files without complaint (including C++ sources) calling appropriate File Handling Interface methods as it does so.

Indeed, if a user creates a new CellML/SED-ML file, then he will be held to that type. That is, unless it's a CellML file and that he uses the `CellML Text` view, makes the file invalid CellML and then decides to save the invalid CellML to a text file (as you have now realised). In that case, he ends up with a pure text file and, indeed, if he opens that text file in OpenCOR, only the `Raw Text` view will be able to render it. The same goes with C++ files or any other text-based file.

However, the key point here is that such text-based files (and the same goes with CellML/SED-ML files) have a file tab associated with them and it is because of that association that some `FileHandlingInterface` methods will be generated for those files.

> How about having the `FileManager` emit signals regardless of a file's type but only call interface methods if the contents are of a known type and valid (for the type)??

Based on what I mentioned about the file tab association, I don't think that would make sense or even be right, unless I misunderstood what you are after.

**Comment by dbrnz** [comment]
Closing since a workspace may contain other files than those created/edited by OpenCOR and we cannot rely upon OpenCOR's notifications to signal modifications to a workspace. Instead periodic polling of the workspace `git status` will be used to detect changes.
